### PR TITLE
First iteration of documentation for API boundary string representation

### LINKED ISSUE #14: String encodings (UTF-8/UTF-16)
We want to make OmnICU able to be ported to other environments where strings are not necessarily UTF-8.  For example, if called via FFI from Dart or WebAssembly, we may want to process strings that are UTF-16.  How should we deal with strings in OmnICU that makes it easy to target UTF-16 environments while not hurting performance in UTF-8 environments?

### PR DESCRIPTION
**Pull request by hsivonen**
Closes #14.

### CODE REVIEW

**Comment by hsivonen** [comment]
Hmm. Maybe I should replace the appender function last resort with an ICU4X-allocated `repr(C)` string object after all.

**Comment by hsivonen** [comment]
Hmm. Or maybe there isn't even a need for a `repr(C)` object but just the buffer taken from a Rust `Vec`/`String` and its length in an outparam. I'll update the doc accordingly.

**Comment by hsivonen** [comment]
Updated with guidance that aims to allocate once in the best case, with example of the case where there's no good worst-case estimation, and with a C++ wrapper sample.

**Comment by sffc** [comment]
Thanks for this!  Looks great overall.

One thought to start: It's common that ICU and 402 functions either accept string input (collation, segmentation, parsing) or return string output (formatting), but not necessarily both.  Although there are cases of functions that take strings as both input and output (you have case mapping as an example), ICU functions frequently maps between strings and machine types (doubles, integers, etc).  Therefore, I'm curious how this affects the memory allocation section where you discuss the need to make a guess at the worst-case output string size.

**Comment by nciric** [comment]
One that takes strings and returns them is message format API (but that one
is the hard on many fronts, including Wasm re-entry).

пет, 24. апр 2020. у 09:56 Shane F. Carr <notifications@github.com> је
написао/ла:

**Comment by sffc** [comment]
I think it would be relevant to this doc to discuss a scenario such as the formatting APIs where the length of the output string is known after some logic is performed.  For example, could ICU4X "ask" the host for a buffer of a certain size via a function pointer?  Is that conventional?

Here's an example:

```rust
type FnAlloc<T> = dyn Fn(usize) -> Box<[T]>;

fn format(_input: f64, alloc: &FnAlloc<u8>) -> Box<[u8]> {
    // Dummy implementation
    let mut result = alloc(3);
    result[0] = 0x41;
    result[1] = 0x42;
    result[2] = 0x43;
    return result;
}

fn alloc_impl(len: usize) -> Box<[u8]> {
    vec![0u8; len].into_boxed_slice()
}

fn main() {
    let result_box = format(1.0, &alloc_impl);
    let result_string = String::from_utf8(result_box.to_vec()).unwrap();
    println!("{}", result_string);
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=eb2248890ed1a746ba3155b7a31c62c9

**Comment by macchiati** [comment]
Often the cost of computing the buffer size ahead of time is a substantial
proportion of the cost of filling the buffer, which makes the total cost
not very efficient. However, it is often not too much work to compute a
maximum size, and report the size actually used afterwards. What about that?

Mark

**Comment by hsivonen** [comment]
>  Although there are cases of functions that take strings as both input and output (you have case mapping as an example), ICU functions frequently map between strings and machine types (doubles, integers, etc).

Per discussion on the call, I'll add clarifying notes that we'd end up using what the document now positions as the last resort case way more often than the document now implies.

> For example, could ICU4X "ask" the host for a buffer of a certain size via a function pointer? Is that conventional?

So ICU4X would call the allocator, but what the allocator is would be caller-customatizable instead of being a matter of customizing the global allocator for Rust code at build time.

Is that use case something that would actually be useful compared to either
1. Using the same allocator for Rust code and the caller, in which case it's OK to use C `free()` to free buffers returned by ICU4X.
2. Building all Rust code in the app with the specific allocator that you want instead of passing its entry point to each ICU4X call.

That is, is there a use case worth supporting where you'd want to have a different allocator for buffers ICU4X and for other Rust code?

> However, it is often not too much work to compute a maximum size, and report the size actually used afterwards. What about that?

This is what the slice-writing APIs would do. Additionally, over the weekend, I figured out how to use Rust `String`/`Vec<u16>` in a way that moves the shrink-to-fit policy to the caller. This would expose both the used length and the capacity to the FFI caller.

I'll add a new changeset to this PR.

**Comment by hsivonen** [comment]
> Additionally, over the weekend, I figured out how to use Rust `String`/`Vec<u16>` in a way that moves the shrink-to-fit policy to the caller.

Unfortunately, the API I had in mind is [nightly-only](https://github.com/rust-lang/rust/issues/65816) still.

**Comment by sffc** [comment]
> So ICU4X would call the allocator, but what the allocator is would be caller-customatizable instead of being a matter of customizing the global allocator for Rust code at build time.
> 
> Is that use case something that would actually be useful compared to either
> 
> 1. Using the same allocator for Rust code and the caller, in which case it's OK to use C `free()` to free buffers returned by ICU4X.
> 2. Building all Rust code in the app with the specific allocator that you want instead of passing its entry point to each ICU4X call.
> 
> That is, is there a use case worth supporting where you'd want to have a different allocator for buffers ICU4X and for other Rust code?

I was mainly thinking about the FFI use case.  This would be a way for the FFI client to allocate a host-side buffer and have ICU4X fill it in with the correct character encoding.  Otherwise, if we allocate the string on the Rust side, it's the host's responsibility to copy the contents out of the buffer to its own buffer.  Maybe that pattern is okay, but I wanted to pose the question.

**Comment by zbraniecki** [comment]
> Unfortunately, the API I had in mind is nightly-only still.

We could try to check with the Rust team if this API is on the path to stabilization. If it is and we can hope to have it in Rust stable within 6 months, I wouldn't mind designing with that in mind.

**Comment by hsivonen** [comment]
> Otherwise, if we allocate the string on the Rust side, it's the host's responsibility to copy the contents out of the buffer to its own buffer. Maybe that pattern is okay, but I wanted to pose the question.

That would still require e.g. a C++ caller to have a custom string class that can adopt a buffer allocated externally to the string class. (As opposed to the string class insisting on managing its buffer strictly on its own.)

If the Rust code is built with the same allocator as the host code, the host code can adopt a buffer that was allocated by ICU4X just as if it came directly from the usual host allocator. If the Rust code is built with a different allocator than the host code, it's still possible to have a custom host string class that adopts the buffer if it can use the ICU4X-provided freeing function and can pass not only the pointer but also the capacity to the freeing function. So I'm skeptical of the value of complicating the ICU4X API by making the allocator selectabel on the call sites.

(Rust supports allocators that, unlike C allocators, aren't capable of looking up the allocation size given a pointer previously returned by the allocator and that instead trust the application to pass back the allocation size. Such an allocator [actually exists for Wasm](https://github.com/rustwasm/wee_alloc).)

> We could try to check with the Rust team if this API is on the path to stabilization.

I checked on the issue, but my check was marked resolved as noise, so you need to expand it to see it. However, by checking, I did learn something that I had missed: while `into_raw_parts` is nightly-only, `from_raw_parts` is not _and_ one can build a function equivalent to `into_raw_parts` from pieces already available on release. I'll update the proposal here accordingly.

(The stabilization of `into_raw_parts` appears to be blocked by two tiny decisions: 1) method or free function and 2) whether the returned pointer is annotated for non-nullness since such annotation is now possible or not annotated for consistency with `from_raw_parts` and `Box::into_raw_parts`.)

**Comment by hsivonen** [comment]
I believe I have now addressed the concerns from last weeks call. Please let me know if I missed something that I should edit in.

File: docs/string-representation.md [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+
+In the Rust context, the name annotation is `_str`. In the FFI context of the input side, the main limitation is `_utf8_unsafe`. The point of the different annotations is that of the Rust side, thanks to the guarantees of Rust, calling these functions is safe. However, when called from the language that doesn't have Rust's guarantees as part of the language, it is up to the programmer to ensure that the input is valid UTF-8. Otherwise, Undefined Behavior ensues. Hence the "unsafe" designation in FFI, i.e. the C API. (This type is irrelevant in the FFI context on the output side.) It is
```
Comment (emilio) [inline docs/string-representation.md:62]: Stray "It is"?

File: docs/string-representation.md [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+
+On the input side, `&str` differs from `&[u8]` holding potentially-invalid UTF-8 in performance. The former can be iterated over by Unicode scalar value without having to have branches that take into account invalid byte sequences. In this sense, the distinction between the function that takes `&str` and a function that takes UTF-8 `&[u8]` is one of performance. Correctness-wise, it is always okay for the function that takes `&str` to call `as_bytes()` on the argument and then delegate to the function that takes `&[u8]`. When performance of the operation is more important than minimizing binary size, the function that takes `&str` should make use of the knowledge that there are no ill-formed sequences.
+
+On the outside, the function that takes `&mut str` always delegates to the version that takes `&[u8]` and trailing loop that zeros bytes after the last byte written by the delegate function until either the end of the slice or UTF a lead byte is reached. This is enough to uphold the invariant of `&mut str`.
```
Comment (emilio) [inline docs/string-representation.md:66]: that takes `&mut [u8]` probably?

File: docs/string-representation.md [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+
+A given ICU4X operation that outputs text must always provide a version, whose input type is `&str` and output type is `&[u8]` (this one gets exposed via FFI as `_unsafe_utf8`), the version whose input type is `&str` and output type is `&mut str`, and a version whose input type is `&str` and that returns a `String` instead of having an output argument. The last two or always implemented as a thin mechanical wrappers around the first one. If performance considerations permit, the first one may just delegate to the version that takes `&[u8]` input. The function that returs a `String` has no name annotation.
```
Comment (emilio) [inline docs/string-representation.md:68]: Output type is `&mut [u8]` in the first one right? But I'd expect the mapping to be: `&[u8]` input -> `&mut [u8]` output (no special name), then `&str` -> `&mut str` (`_unsafe_utf8` in the FFI), and `&str` -> `String`, right?

File: docs/string-representation.md [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+    vec.resize(src.len(), 0);
+    let (read_first, written_first) = to_lowercase_str_utf8(src, &mut vec);
+    let mut total_written = written_first;
+    let input_left = src.len() - read_first;
+    if input_left != 0 {
+        vec.resize(
+            written_first + to_lowercase_utf8_max(input_left).unwrap(),
+            0,
+        );
+        let (read_second, written_second) = to_lowercase_str_utf8(src, &mut vec);
```
Comment (emilio) [inline docs/string-representation.md:264]: This should be `to_lowercase_str_utf8(&src[read_first..], &mut vec[written_first..])` right?

File: docs/string-representation.md (hunk 3 above)
Comment (hsivonen) [reply docs/string-representation.md:68]: `_unsafe_utf8` in FFI reads from `&str` and writes to `&mut [u8]`.

File: docs/string-representation.md (hunk 1 above)
Comment (hsivonen) [reply docs/string-representation.md:62]: Thanks. Will be fixed in next push.

File: docs/string-representation.md (hunk 2 above)
Comment (hsivonen) [reply docs/string-representation.md:66]: Thanks.

File: docs/string-representation.md (hunk 3 above)
Comment (hsivonen) [reply docs/string-representation.md:68]: Fixed.

File: docs/string-representation.md (hunk 4 above)
Comment (hsivonen) [reply docs/string-representation.md:264]: Whoa. Yes. Thanks.

**Pull request merged**